scipy/linalg/tests/test_decomp.py

```diff
@@ -1004,128 +1011,169 @@ def test_empty(self, dt):
+        atol = 1e-7 if dtype == "float64" else 1e-5
+        u, s, vh = svd(a, lapack_driver=self.lapack_driver)
+        xp_assert_close(u.T @ u, xp.eye(3, dtype=getattr(xp, dtype)), atol=atol)
+        xp_assert_close(vh.T @ vh, xp.eye(3, dtype=getattr(xp, dtype)), atol=atol)
+        sigma = xp.zeros((u.shape[0], vh.shape[0]), dtype=s.dtype)
+        for i in range(s.shape[0]):
+            sigma[i, i] = s[i]
+        xp_assert_close(u @ sigma @ vh, a, atol=atol)
+
+    @skip_xp_backends(np_only=True)
+    @pytest.mark.parametrize("dtype", [np.int32, np.int64])
+    def test_dtypes_nonstandard(self, dtype, xp):
+        a = np.asarray([[1, 2, 3], [1, 20, 3], [2, 5, 6]], dtype=dtype)
+        atol = 1e-7
+        u, s, vh = svd(a, lapack_driver=self.lapack_driver)
+        xp_assert_close(u.T @ u, np.eye(3), atol=atol)
+        xp_assert_close(vh.T @ vh, np.eye(3), atol=atol)
+        sigma = np.zeros((u.shape[0], vh.shape[0]), dtype=s.dtype)
+        for i in range(s.shape[0]):
+            sigma[i, i] = s[i]
+        xp_assert_close(u @ sigma @ vh, a.astype(np.float64))
+    
+    def test_simple(self, xp):
+        a = xp.asarray([[1., 2, 3], [1, 20, 3], [2, 5, 6]])
         for full_matrices in (True, False):
             u, s, vh = svd(a, full_matrices=full_matrices,
                            lapack_driver=self.lapack_driver)
-            assert_array_almost_equal(u.T @ u, eye(3))
-            assert_array_almost_equal(vh.T @ vh, eye(3))
-            sigma = zeros((u.shape[0], vh.shape[0]), s.dtype.char)
-            for i in range(len(s)):
+            xp_assert_close(u.T @ u, xp.eye(3), atol=1e-6)
+            xp_assert_close(vh.T @ vh, xp.eye(3), atol=1e-6)
+            sigma = xp.zeros((u.shape[0], vh.shape[0]), dtype=s.dtype)
+            for i in range(s.shape[0]):
                 sigma[i, i] = s[i]
-            assert_array_almost_equal(u @ sigma @ vh, a)
+            xp_assert_close(u @ sigma @ vh, a, rtol=1e-6)
 
-    def test_simple_singular(self):
-        a = [[1, 2, 3], [1, 2, 3], [2, 5, 6]]
+    def test_simple_singular(self, xp):
+        a = xp.asarray([[1., 2, 3], [1, 2, 3], [2, 5, 6]])
         for full_matrices in (True, False):
             u, s, vh = svd(a, full_matrices=full_matrices,
                            lapack_driver=self.lapack_driver)
-            assert_array_almost_equal(u.T @ u, eye(3))
-            assert_array_almost_equal(vh.T @ vh, eye(3))
-            sigma = zeros((u.shape[0], vh.shape[0]), s.dtype.char)
-            for i in range(len(s)):
+            xp_assert_close(u.T @ u, xp.eye(3), atol=1e-6)
+            xp_assert_close(vh.T @ vh, xp.eye(3), atol=1e-6)
+            sigma = xp.zeros((u.shape[0], vh.shape[0]), dtype=s.dtype)
+            for i in range(s.shape[0]):
                 sigma[i, i] = s[i]
-            assert_array_almost_equal(u @ sigma @ vh, a)
+            xp_assert_close(u @ sigma @ vh, a, rtol=1e-6)
```

Comment:
the tolerances around here could do with a look, I'm not sure why I wrote a mixture of `atol` and `rtol`.

Reply:
ah I remember, it's because the default for `np.testing.assert_array_almost_equal` is roughly equivalent to `rtol=0, atol=1.5e-6`.